### ISSUE TITLE
Potential issue with stiffness matrix (K44 and K55)

### ISSUE THREAD

**Issue by sdillenburg** [capytaine/capytaine#178]
Dear all,

The horizontal (x and y) component of the center of gravity refers to the (local) coordinates of the center of the body, the vertical component, however, is defined relative to the global z = 0. See following example:

The function `FloatingBody.center_of_buoyancy ` returns the center of buoyancy in global coordinates. Following line suggests the center of gravity is also defined in global coordinates:

https://github.com/mancellin/capytaine/blob/8a07364afa8092cc30f1aefed57fab16b5c96034/docs/user_manual/examples/boat_animation.py#L18

However, when a body has its center off x = 0 and y = 0, results become strange, see following test:

```
import capytaine as cpt
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt

# Define the bodies
# Mass matrix
m = 4612500 # kg
radii = np.array([9,9,12]) # m

M = np.eye(6)*m
M[3:,3:] *= radii**2

# Bodies
barge_0 = cpt.RectangularParallelepiped(size=(30,30,5), resolution=(20,20,5), center=(0,-20,-2.5),top=False,name="barge_0")
barge_0.center_of_mass = (0,0,5)
barge_0.add_all_rigid_body_dofs()
barge_0.hydrostatic_stiffness = barge_0.compute_hydrostatic_stiffness(rho=1025)
barge_0.inertia_matrix = xr.DataArray(M, dims=['influenced_dof', 'radiating_dof'],
                                         coords={'influenced_dof': list(barge_0.dofs),'radiating_dof': list(barge_0.dofs)},
                                         name='inertia_matrix')
                                         
barge_1 = cpt.RectangularParallelepiped(size=(30,30,5), resolution=(20,20,5), center=(0,-20,-2.5),top=False,name="barge_1")
barge_1.center_of_mass = (0,-20,5)
barge_1.add_all_rigid_body_dofs()
barge_1.hydrostatic_stiffness = barge_1.compute_hydrostatic_stiffness(rho=1025)
barge_1.inertia_matrix = xr.DataArray(M, dims=['influenced_dof', 'radiating_dof'],
                                         coords={'influenced_dof': list(barge_1.dofs),'radiating_dof': list(barge_1.dofs)},
                                         name='inertia_matrix')

print(barge_0.center_of_buoyancy)
print(barge_1.center_of_buoyancy)

problems_0 = []
problems_1 = []

omega_range = np.linspace(0.15, 2, 25)
direction_range = [np.pi/2] #np.linspace(0, 2*np.pi, 7)

for omega in omega_range:
     for dof in list(barge_0.dofs):
        problems_0.append(cpt.RadiationProblem(body=barge_0, radiating_dof=dof, omega=omega, rho=1025))
        for wave_direction in direction_range: 
            problems_0.append(cpt.DiffractionProblem(body=barge_0, omega=omega, wave_direction=wave_direction,rho=1025))

     for dof in list(barge_1.dofs):
        problems_1.append(cpt.RadiationProblem(body=barge_1, radiating_dof=dof, omega=omega, rho=1025))
        for wave_direction in direction_range:
            problems_1.append(cpt.DiffractionProblem(body=barge_1, omega=omega, wave_direction=wave_direction,rho=1025))

# Solves the problem
solver = cpt.BEMSolver()
results_0 = solver.solve_all(problems_0,n_jobs=4)
data_0 = cpt.assemble_dataset(results_0)

results_1 = solver.solve_all(problems_1,n_jobs=4)
data_1 = cpt.assemble_dataset(results_1)

data_0['rao090'] = cpt.post_pro.rao(data_0, wave_direction=np.pi/2)
data_1['rao090'] = cpt.post_pro.rao(data_1, wave_direction=np.pi/2)

fig, ax = plt.subplots()

ax.plot(omega_range, np.abs(data_0['rao090'].sel(radiating_dof='Heave')),marker='o',label='COG in coords wrt to body center')
ax.plot(omega_range, np.abs(data_1['rao090'].sel(radiating_dof='Heave')),marker='o',label='COG in global coords')

plt.legend()
plt.tight_layout()
plt.savefig('raos_cog_test.svg')
```

![raos_cog_test](https://user-images.githubusercontent.com/45045610/180172797-07a7ac35-4b4a-4389-bb27-022e8c413103.svg)

The vertical coordinate for the center of gravity definition still seems to be 0, see
https://github.com/mancellin/capytaine/blob/8a07364afa8092cc30f1aefed57fab16b5c96034/capytaine/bodies/bodies.py#L316

I suggest to update the documentation stating the reference frame for definition of the center of gravity.

**Comment by mancellin** [comment]
In principle, all coordinates are in the global frame, there is no local frame.

Indeed, there is something suspicious with the hydrostatic stiffness.
Here is a simplified version of your example:
```python
import numpy as np
import capytaine as cpt

barge_0 = cpt.RectangularParallelepiped(
        size=(30.0, 30.0, 5.0), resolution=(20, 20, 5),
        center=(0.0, 0.0, -2.5), top=False, name="barge_0",
      )
barge_0.center_of_mass = (0.0, 0.0, 5.0)
barge_0.add_all_rigid_body_dofs()

barge_1 = cpt.RectangularParallelepiped(
        size=(30.0, 30.0, 5.0), resolution=(20, 20, 5),
        center=(0.0, -20.0, -2.5), top=False, name="barge_1",
      )
barge_1.center_of_mass = (0.0, -20.0, 5.0)
barge_1.add_all_rigid_body_dofs()

hs0 = barge_0.compute_hydrostatic_stiffness()
hs1 = barge_1.compute_hydrostatic_stiffness()

print(np.max(np.abs(hs0 - hs1)))
```
I would expect the difference to be zero, but it is not.

**Comment by sdillenburg** [comment]
Right. The heave-roll and heave-pitch coupling seems to be affected.

**Comment by sdillenburg** [comment]
Heave-roll and heave-pitch terms seem to be affected. Maybe the center of the body needs to be subtracted from $x$ and $y$ in: 
$K_{34} = -\rho g \iint_S y n_3 \\, dS$
$K_{35} = -\rho g \iint_S x n_3 \\, dS$

I will have a look at the source later.

**Comment by sdillenburg** [comment]
I did some tests and checks. It seems there is an issue with K44 and K55, the off-diagonal terms match. The diagonal should be the same for both positions, this is not the case in the matrices computed by capytaine.
Please have a look at the following example:

```
import capytaine as cpt
import numpy as np
import xarray as xr

# Define box shaped bodies
length = 30 # m
beam = 30 # m
draft = 5 # m

# Offset of barge_1 (barge_0 remains at x = 0, y = 0)
barge_1_x = 10 # m
barge_1_y = 20 # m

# Constants
rho = 1025 # kg/m**3
g = 9.81 # m/s**2

# Mass matrix
zcg = 0 # m 
m = length*beam*draft*rho # kg
radii = np.array([np.sqrt((beam**2+draft**2)/12),np.sqrt((length**2+draft**2)/12),np.sqrt((length**2+beam**2)/12)]) # m

M = np.eye(6)*m # kg
M[3:,3:] *= radii**2 # kg*m**2

# Stiffness matrix
Awl = length*beam # m**2
KB = 0.5*draft # m
KG = zcg # m
BMT = beam**2/12/draft # m
BML = length**2/12/draft # m
GMT = KB+BMT-KG # m
GML = KB+BML-KG # m

print('Mass: {} kg'.format(m))
print('GMT: {} m'.format(GMT))
print('GML: {} m'.format(GML))

K_0 = np.array([[0., 0.,  0.,        0.,      0.,      0.],
                [0., 0.,  0.,        0.,      0.,      0.],
                [0., 0.,  rho*g*Awl, 0,       0,       0.],
                [0., 0.,  0,         m*g*GMT, 0.,      0.],
                [0., 0.,  0,         0.,      m*g*GML, 0.],
                [0., 0.,  0.,        0.,      0.,      0.]])

K_1 = np.array([[0., 0.,  0.,                   0.,                             0.,                            0.],
                [0., 0.,  0.,                   0.,                             0.,                            0.],
                [0., 0.,  rho*g*Awl,            rho*g*Awl*barge_1_y,           -rho*g*Awl*barge_1_x,           0.],
                [0., 0.,  rho*g*Awl*barge_1_y,  m*g*GMT,                       -rho*g*Awl*barge_1_x*barge_1_y, 0.],
                [0., 0., -rho*g*Awl*barge_1_x, -rho*g*Awl*barge_1_x*barge_1_y,  m*g*GML,                       0.],
                [0., 0.,  0.,                   0.,                             0.,                            0.]])
#
# Body definition
barge_0 = cpt.RectangularParallelepiped(size=(length,beam,draft), resolution=(int(length/1.5),int(beam/1.5),draft), center=(0,0,-0.5*draft),top=False,name='barge_0')
barge_0.center_of_mass = (0,0,zcg)
barge_0.add_all_rigid_body_dofs()
barge_0.hydrostatic_stiffness = xr.DataArray(K_0, dims=['influenced_dof', 'radiating_dof'],
                                                  coords={'influenced_dof': list(barge_0.dofs),'radiating_dof': list(barge_0.dofs)},
                                                  name='hydrostatic_stiffness_matrix')
barge_0.inertia_matrix = xr.DataArray(M, dims=['influenced_dof', 'radiating_dof'],
                                         coords={'influenced_dof': list(barge_0.dofs),'radiating_dof': list(barge_0.dofs)},
                                         name='inertia_matrix')
                                         
barge_1 = cpt.RectangularParallelepiped(size=(length,beam,draft), resolution=(int(length/1.5),int(beam/1.5),draft), center=(barge_1_x,barge_1_y,-0.5*draft),top=False,name='barge_1')
barge_1.center_of_mass = (barge_1_x,barge_1_y,zcg)
barge_1.add_all_rigid_body_dofs()
barge_1.hydrostatic_stiffness = xr.DataArray(K_1, dims=['influenced_dof', 'radiating_dof'],
                                                  coords={'influenced_dof': list(barge_1.dofs),'radiating_dof': list(barge_1.dofs)},
                                                  name='hydrostatic_stiffness_matrix')
barge_1.inertia_matrix = xr.DataArray(M, dims=['influenced_dof', 'radiating_dof'],
                                         coords={'influenced_dof': list(barge_1.dofs),'radiating_dof': list(barge_1.dofs)},
                                         name='inertia_matrix')

# Compare hydrostatic stiffness matrices
K_0_cpt = barge_0.compute_hydrostatic_stiffness(rho=rho)
K_1_cpt = barge_1.compute_hydrostatic_stiffness(rho=rho)

np.set_printoptions(precision=3)
diff_K_0 = K_0-K_0_cpt.values
diff_K_1 = K_1-K_1_cpt.values

print('\nDifference Capytaine vs. Handcheck')
print(diff_K_0)
print(diff_K_1)

print(diff_K_0[abs(diff_K_0)>1])
print(diff_K_1[abs(diff_K_1)>1])

print('\nDifference barge_0 vs. barge_1 (Capytaine)')
print(K_0_cpt.values-K_1_cpt.values)
```

I have double checked the GM values with [MOSES](https://www.bentley.com/de/products/product-line/offshore-structural-analysis-software/moses).

@[RubendeBruin](https://github.com/mancellin/capytaine/issues?q=is%3Aissue+is%3Aopen+author%3ARubendeBruin), could you maybe check with DAVE also? Just to make sure I am not missing anything...

**Comment by sdillenburg** [comment]
I accidentally closed this issue. Re-opening it...

**Comment by mancellin** [comment]
The issue could come from #166. To handle non-neutrally buoyant bodies, @salhus and I used formulas from WAMIT manual which might have been written for a center of mass expressed in local coordinates.

**Comment by mancellin** [comment]
Most likely, the root of the issue is that for rigid body `compute_hydrostatic_stiffness` assumes the rotation dofs are defined around `(0, 0, 0)` even when they have been defined around the center of mass.

**Comment by salhus** [comment]
@mancellin 
I think you are right about the hydrodynamics being calculated about the CoG. I had a discussion with the WEC-Sim Team, and the reason WEC-Sim doesn't have those extra terms in K44 and K55 is because hydrodynamics coefficients need to be calculated about the CoG for Simscape Multibody. It is easier to model multi-body dynamics about the CoG and also helps establish a standard approach for all bodies (submerged or otherwise).
However, for a BEM solver the ability to generate hydrodynamics coefficients with respect to any point will require those extra terms to be present in the formulation. The choice of reference for modeling the dynamics can then be made by the user.

**Comment by mancellin** [comment]
This should have been fixed by #196.
I'll release a v1.4.1 with this patch soon.
If you have time to test the master branch test beforehand, it would be appreciated.
